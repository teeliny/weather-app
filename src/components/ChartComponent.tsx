import React from 'react';
import {
  Chart as BarChart,
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
} from 'chart.js';
import { Bar } from 'react-chartjs-2';
import { convertTemp, timeFormats } from '../utils/basicFormatter';

interface IChartComp {
  temp: number;
  humidity: number;
  wind_speed: number;
  cloud: string;
  current_date: string;
  hour: number;
}
interface IChartInput {
  input: IChartComp[];
  tempUnit: string;
}

BarChart.register(
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
);

function ChartComponent({ input, tempUnit }: IChartInput) {
  const currentDate = input[0].current_date;
  const barData = [];
  const barLabel = [];
  for (let index = 0; index < input.length; index++) {
    const { temp, hour } = input[index];
    const hourString: string = hour.toString();
    barData.push((convertTemp(tempUnit, temp)).toFixed(2));
    barLabel.push(timeFormats[hourString]);
  }
  const options = {
    responsive: true,
    plugins: {
      title: {
        display: true,
        text: `Chart for ${currentDate} in ${tempUnit === '0' ? 'C' : 'F'}`,
      },
    },
  };
  const data = {
    labels: barLabel,
    datasets: [
      {
        data: barData,
        backgroundColor: 'rgba(255, 99, 132, 0.5)',
      }
    ],
  }; 
  return (
    <React.Fragment>
      <Bar options={options} data={data} />
    </React.Fragment>
  )
}

export default ChartComponent;

