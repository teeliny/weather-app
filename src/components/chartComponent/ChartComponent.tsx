import React from 'react';
import {
  Chart as BarChart,
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
} from 'chart.js';
import { Bar } from 'react-chartjs-2';
import { convertTemp, timeFormats } from '../../utils/formatter/basicFormatter';
import { IChartInput } from './chart.typing';


BarChart.register(
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
);

function ChartComponent({ input, tempUnit }: IChartInput) {
  const currentDate = input[0].current_date;
  const barData = [];
  const barLabel = [];
  for (let index = 0; index < input.length; index++) {
    const { temp, hour } = input[index];
    const hourString: string = hour.toString();
    barData.push((convertTemp(tempUnit, temp)).toFixed(2));
    barLabel.push(timeFormats[hourString]);
  }
  const options = {
    maintainAspectRatio: false,
    responsive: true,
    scales: {
      x: {
        grid: {
          color: '#4eeb9c',
          tickColor: '#0055ff',
        },
        ticks: {
          color: '#0055ff',
        },
      },
      y: {
        grid: {
          color: '#4eeb9c',
          tickColor: '#0055ff',
        },
        ticks: {
          color: '#0055ff',
        },
      },
    },
    plugins: {
      title: {
        display: true,
        text: `Chart for ${currentDate} in ${tempUnit === '0' ? 'C' : 'F'}`,
        color: '#0055ff',
      },
    },
  };
  const data = {
    labels: barLabel,
    datasets: [
      {
        data: barData,
        backgroundColor: 'rgba(255, 99, 132, 0.5)',
      }
    ],
  }; 
  return (
    <React.Fragment>
      <Bar
        options={options}
        data={data}
        height={280}
      />
    </React.Fragment>
  )
}

export default ChartComponent;

