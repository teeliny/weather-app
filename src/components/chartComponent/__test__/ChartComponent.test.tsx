import React from 'react';
import { render, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ChartComponent from '../ChartComponent';

let eleWrapper: JSX.Element;

beforeEach(() => {
  eleWrapper = (
    <ChartComponent
      input={[
        {
          temp: 30,
          humidity: 60,
          wind_speed: 2.54,
          cloud: 'cloudy',
          hour: 3,
          current_date: '12 Dec. 2021'
        },
        {
          temp: 30.5,
          humidity: 75,
          wind_speed: 2.08,
          cloud: 'mist cloud',
          hour: 6,
          current_date: '12 Dec. 2021'
        }
      ]}
      tempUnit={'0'}
    />
  )
});

afterEach(cleanup);

it('render canvas in chart component correctly', () => {
  const { container } = render(eleWrapper);
  expect(container.querySelector('canvas')).toBeInTheDocument();
})

it('matches bar chart snapshot', () => {
  const tree = renderer.create(eleWrapper).toJSON();
  expect(tree).toMatchSnapshot();
})