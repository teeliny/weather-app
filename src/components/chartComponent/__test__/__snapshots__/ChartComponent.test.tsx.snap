// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches bar chart snapshot 1`] = `
<canvas
  height={280}
  role="img"
  width={300}
/>
`;
